### ISSUE TITLE
Prevent cascading diagnostics when using unknown type form

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ty#667]
We currently report multiple diagnostics in situations like the following:

```py
# Name `UnknownTypeForm` used when not defined (unresolved-reference) [Ln 1, Col 4]
# Int literals are not allowed in this context in a type expression (invalid-type-form) [Ln 1, Col 20]
x: UnknownTypeForm[1]


# Name `UnknownTypeForm` used when not defined (unresolved-reference) [Ln 1, Col 4]
# Syntax error in forward annotation: Unexpected token at the end of an expression (invalid-syntax-in-forward-annotation) [Ln 1, Col 20]
x: UnknownTypeForm["some string"]
```

The first of these (`unresolved-reference`) is obviously fine, but I think we should suppress the other diagnostic. This can also cause false positives in unreachable code where legit type forms may be inferred as `Unknown`:

```py
assert False

from typing_extensions import Literal

def outer():
    # Syntax error in forward annotation: Unexpected token at the end of an expression (invalid-syntax-in-forward-annotation) [Ln 7, Col 26]
    def inner(p: Literal["missing value"]):
        pass
```

**Comment by InSyncWithFoo** [comment]
A possible approach is to apply "name-based resolution" on some of the unresolved references:

```python
# `Literal` is unresolved, since it is not imported,
# but instead of treating it as `Unknown`,
# assume the user mean `typing[_extensions].Literal`.

#        vvvvvvv error: `Literal` is not resolved
def _(a: Literal['a b']):
#                ^^^^^ Fine
	reveal_type(a)        # Literal['a b']
	reveal_type(Literal)  # Error out, but reveal `typing[_extensions].Literal`
```

`reveal_type()` is already benefitting from this special treatment. I'm not sure how much of a good idea it is to extend to other symbols, though.

**Comment by carljm** [comment]
I don't think we want to make assumptions based on the name.

It seems to me the likely fix here is just some kind of "visit and evaluate/store unknown" helper in type expression evaluation that doesn't emit diagnostics, so when we evaluate something as `Unknown`, we can just mark its subscript expression (and all sub-expressions) as also `Unknown` and move on, rather than really trying to evaluate them and possibly emitting diagnostics.

**Comment by MeGaGiGaGon** [comment]
Looks like this was fixed in `0.0.1-alpha.22` and `0.0.1-alpha.29`:
```powershell
PS ~>echo @'
# Name `UnknownTypeForm` used when not defined (unresolved-reference) [Ln 1, Col 4]
# Int literals are not allowed in this context in a type expression (invalid-type-form) [Ln 1, Col 20]
x: UnknownTypeForm[1]


# Name `UnknownTypeForm` used when not defined (unresolved-reference) [Ln 1, Col 4]
# Syntax error in forward annotation: Unexpected token at the end of an expression (invalid-syntax-in-forward-annotation) [Ln 1, Col 20]
x: UnknownTypeForm["some string"]

assert False

from typing_extensions import Literal

def outer():
    # Syntax error in forward annotation: Unexpected token at the end of an expression (invalid-syntax-in-forward-annotation) [Ln 7, Col 26]
    def inner(p: Literal["missing value"]):
        pass
'@ > issue.py

PS ~>uvx ty@0.0.1-alpha.21 check issue.py --output-format concise
issue.py:3:4: error[unresolved-reference] Name `UnknownTypeForm` used when not defined
issue.py:3:20: error[invalid-type-form] Int literals are not allowed in this context in a type expression
issue.py:8:4: error[unresolved-reference] Name `UnknownTypeForm` used when not defined
issue.py:8:20: error[invalid-syntax-in-forward-annotation] Syntax error in forward annotation: Unexpected token at the end of an expression
issue.py:16:26: error[invalid-syntax-in-forward-annotation] Syntax error in forward annotation: Unexpected token at the end of an expression
Found 5 diagnostics

PS ~>uvx ty@0.0.1-alpha.22 check issue.py --output-format concise
issue.py:3:4: error[unresolved-reference] Name `UnknownTypeForm` used when not defined
issue.py:3:20: error[invalid-type-form] Int literals are not allowed in this context in a type expression
issue.py:8:4: error[unresolved-reference] Name `UnknownTypeForm` used when not defined
issue.py:8:20: error[invalid-syntax-in-forward-annotation] Syntax error in forward annotation: Unexpected token at the end of an expression
Found 4 diagnostics

PS ~>uvx ty@0.0.1-alpha.29 check issue.py --output-format concise
issue.py:3:4: error[unresolved-reference] Name `UnknownTypeForm` used when not defined
issue.py:8:4: error[unresolved-reference] Name `UnknownTypeForm` used when not defined
Found 2 diagnostics
```
